pandas/core/generic.py

```diff
@@ -9020,22 +9020,22 @@ def tshift(
         if isinstance(freq, str):
             freq = to_offset(freq)
 
-        block_axis = self._get_block_manager_axis(axis)
+        axis = self._get_axis_number(axis)
         if isinstance(index, PeriodIndex):
             orig_freq = to_offset(index.freq)
-            if freq == orig_freq:
-                new_data = self._data.copy()
-                new_data.axes[block_axis] = index.shift(periods)
-            elif orig_freq is not None:
+            if freq != orig_freq:
+                assert orig_freq is not None  # for mypy
                 raise ValueError(
                     f"Given freq {freq.rule_code} does not match "
                     f"PeriodIndex freq {orig_freq.rule_code}"
                 )
+            new_ax = index.shift(periods)
         else:
-            new_data = self._data.copy()
-            new_data.axes[block_axis] = index.shift(periods, freq)
+            new_ax = index.shift(periods, freq)
 
-        return self._constructor(new_data).__finalize__(self)
+        result = self.copy()
```

Comment:
we are *very* inconsistent on how calling .set_axis in generic.py, e.g. whether using inplace True/False & also _reset_cached_item.

ok here, but think this would be good to be consistent about.

Reply:
yah, #31784 has got me looking at the caching thing now.  is this just to avoid the perf overhead of creating Series objects?

Reply:
yes, it *may* be possible to rip this out as it causes a number of bugs, but would have to see perf impacts.